### DISCUSSION TITLE
Nova reads a Policy multiple times - slows down the page

### DISCUSSION THREAD

**Discussion by yumidevelopment** [laravel/nova-issues#5234] [Q&A] [answered]
- Laravel Version: 9.x
- Nova Version: 4.x
- PHP Version: 8.x

### Description:
We have a nova installation running in production with over 50 different Models/Resources and correspoding Policies. What we are noticing is that the page is running very slow, most likely because there are many duplicate calls to the Policies. In our case over 250 calls and memory is increasing for every Policy we add.

If we remove all the Policies, page is running fast again. I went and created a fresh laravel + nova installation, adding only one model and policy and found out that there are many duplicate calls to e.g. `viewAny`, `view `and `create`. Am I perhaps doing something wrong?

### Detailed steps to reproduce the issue on a fresh Nova installation:
- Create a fresh laravel installation
- Install Nova as described here https://nova.laravel.com/docs/4.0/installation.html
- `php artisan make:model Feed --policy`
- `php artisan nova:resource Feed`
- Add `return true;` in every method in `FeedPolicy`
- Refresh your nova dashboard and see 10 Gate calls

![image](https://user-images.githubusercontent.com/109904203/212261193-eca61232-ab65-43c7-a983-b50297164c8f.png)

**Comment by crynobone** [comment, marked as answer]
This look correct, An index view would need to verify if user can:

-  view detail
- create new record
- edit each particular record
- replicate each particular record
- delete each particular record

**Reply by yumidevelopment** [reply]
But then why does it make so many duplicate calls? I'm quite sure that, because we have so many different policies, the duplicate calls are what is slowing down our page so much. This makes me question if we can use the default Laravel policies at all. Is it a best practise to use them, or to not use them?

**Reply by timboozle** [reply]
We have around 35 active models + policies and for every page change Debugbar is reporting 250+ Gate checks. With them enabled the memory usage reported in Debugbar is 25mb, while disabled shows a memory usage of 2mb. With policies disabled, Nova is much snappier to navigate. Is there a way around this?

**Reply by davidhemphill** [reply]
@timboozle Do your policies hit the database? If you disable policy checks that hit the database, it would make sense that the memory usage will go down.

**Reply by timboozle** [reply]
@davidhemphill Yes you are right, thanks. Replaced all the database checks with `return true;` and we're back to snappy Nova. Guess I got thrown off by the high number of gate checks. I will cache the user's permissions to solve this!

**Reply by shahabzebare** [reply]
@davidhemphill I see this problem in my application, I have one application that have more than 140 Resource with Policy that we have 500+ permission, Loading page with Inertia was some time take 5 to 15 S, this was make my app very slow 
I solved this problem, by creating some package make lazy loading resource based on page now my application max time is 300ms,
some thing nova deal with them make application slow when scale application size not users I mean policess. 
I will publish package or if you want i will make a contribution on laravel nova repository.
I don't know any one deal with nova with strong access control it face this slow problem .

**Reply by crynobone** [reply]
@shahabzebare feel free to release a package.

**Reply by webard** [reply]
@shahabzebare did you try to use once() in authorization methods? I'm curious how you solve this problem, waiting for package!

**Reply by iamx** [reply]
```
/**
 * Determine if the given resource is authorizable.
 *
 * @return bool
 */
public static function authorizable()
{
    return false;
}
```

You could skip the whole authorization of a resource with above. Not sure if that would fit your case or if anyone sees any drawback with this.

**Reply by Andrewrafat** [reply]
<3

**Reply by heshammostafa123** [reply]
@shahabzebare
I have a problem with the speed can you release your package

**Comment by Andrewrafat** [comment]
💗

**Reply by AbdallahAli912** [reply]
where is the answer ? i see no solution here.
@crynobone

**Comment by shahabzebare** [comment]
I published this package it give good supercharge to nova application
-----
https://github.com/shahabzebare/laravel-nova-turbo
